src/iguana/algorithms/clas12/SectorFinder/Algorithm.cc

```diff
@@ -9,96 +9,165 @@ namespace iguana::clas12 {
+        b_track           = GetBankIndex(banks, "REC::Track");
+        b_scint           = GetBankIndex(banks, "REC::Scintillator");
+        setDefaultBanks = true;
+      }
+      userSpecifiedBank_uncharged = false;
     }
 
     // create the output bank
     // FIXME: generalize the groupid and itemid
-    auto result_schema = CreateBank(banks, b_result, "REC::Particle::Sector", {"sector/I"}, 0xF000, 2);
+    auto result_schema = CreateBank(banks, b_result, "REC::Particle::Sector", {"sector/I","pindex/I"}, 0xF000, 4);
     i_sector = result_schema.getEntryOrder("sector");
+    i_pindex = result_schema.getEntryOrder("pindex");
   }
 
-
   void SectorFinder::Run(hipo::banklist& banks) const
   {
     auto& particleBank = GetBank(banks, b_particle, "REC::Particle");
     auto& resultBank   = GetBank(banks, b_result, "REC::Particle::Sector");
 
+    std::vector<int> sectors_uncharged;
+    std::vector<int> pindices_uncharged;
+    if(userSpecifiedBank_uncharged){
+      auto const& userBank = GetBank(banks, b_user_uncharged);
+      GetListsSectorPindex(userBank,sectors_uncharged,pindices_uncharged);
+    }
+
+    std::vector<int> sectors_charged;
+    std::vector<int> pindices_charged;
+    if(userSpecifiedBank_charged){
+      auto const& userBank = GetBank(banks, b_user_charged);
+      GetListsSectorPindex(userBank,sectors_charged,pindices_charged);
+    }
+
+    std::vector<int> sectors_track;
+    std::vector<int> pindices_track;
+    if(!userSpecifiedBank_charged || !userSpecifiedBank_uncharged){
+      auto const& trackBank = GetBank(banks, b_track);
+      GetListsSectorPindex(trackBank,sectors_track,pindices_track);
+    }
+
+    std::vector<int> sectors_cal;
+    std::vector<int> pindices_cal;
+    if(!userSpecifiedBank_charged || !userSpecifiedBank_uncharged){
+      auto const& calBank = GetBank(banks, b_calorimeter);
+      GetListsSectorPindex(calBank,sectors_cal,pindices_cal);
+    }
+
+    std::vector<int> sectors_scint;
+    std::vector<int> pindices_scint;
+    if(!userSpecifiedBank_charged || !userSpecifiedBank_uncharged){
+      auto const& scintBank = GetBank(banks, b_scint);
+      GetListsSectorPindex(scintBank,sectors_scint,pindices_scint);
+    }
+    
+    
     // sync new bank with particle bank, and fill it with zeroes
     resultBank.setRows(particleBank.getRows());
     resultBank.getMutableRowList().setList(particleBank.getRowList());
-    for(int row = 0; row < resultBank.getRows(); row++)
+    for(int row = 0; row < resultBank.getRows(); row++){
       resultBank.putInt(i_sector, row, 0);
-
-    // filter the input bank for requested PDG code(s)
-    if(userSpecifiedBank) { // if user specified a specific bank
-      auto const& userBank = GetBank(banks, b_user);
-      for(auto const& row : particleBank.getRowList()) {
-        if(userBank.getRowList().size() > 0) {
-          resultBank.putInt(i_sector, row, GetSector(userBank, row));
-        }
-      }
+      resultBank.putInt(i_pindex, row, row);
     }
```

Comment:
Should the default "no sector" value be `0` or `-1` (to match the "no sector" return value of the action function)?

If `-1` is preferred, be sure to update the "no sector" cut in the validator.

Reply:
I think 0 is preferable because hipo files have sector = 0 for CD detectors that don't have sectors (the CND has non zero values of sector but this then refers to something else).